src/graph/LookupExecutor.cpp

```diff
@@ -774,27 +803,61 @@ bool LookupExecutor::processFinalEdgeResult(RpcResponse &rpcResp, const Callback
-                std::for_each(returnCols_.begin(), returnCols_.end(), [&](auto& col) {
-                    colTypes.emplace_back(schema->getFieldType(col).get_type());
-                });
+                // when the columns really requested to client (in lookup())are empty,
+                // resp.get_schema() will be nullptr
+                if (resp.get_schema() != nullptr) {
+                    schema = std::make_shared<ResultSchemaProvider>(*(resp.get_schema()));
+                }
+                for (auto& col : returnCols_) {
+                    if (col == _SRC || col == _DST) {
+                        colTypes.emplace_back(nebula::cpp2::SupportedType::VID);
+                    } else {
+                        if (schema == nullptr) {
+                            LOG(ERROR) << "Schema is null unexpected, "
+                                          "get fieldType failed for " << col;
+                            return false;
+                        }
+                        colTypes.emplace_back(schema->getFieldType(col).get_type());
+                    }
+                }
             }
         }
         for (const auto& data : *resp.get_edges()) {
             const auto& edge = data.get_key();
             record.emplace_back(edge.get_src());
             record.emplace_back(edge.get_dst());
             record.emplace_back(edge.get_ranking());
+            yieldRecord.clear();
             for (auto& column : returnCols_) {
-                auto reader = RowReader::getRowReader(data.get_props(), schema);
-                auto res = RowReader::getPropByName(reader.get(), column);
-                if (!ok(res)) {
-                    LOG(ERROR) << "Skip the bad value for prop " << column;
-                    continue;
+                VariantType v;
+                if (column == _SRC) {
+                    v = edge.get_src();
+                } else if (column == _DST) {
+                    v = edge.get_dst();
+                } else {
+                    if (schema == nullptr) {
+                        LOG(ERROR) << "Schema is null unexpected, get value failed for " << column;
+                        return false;
+                    }
+                    auto reader = RowReader::getRowReader(data.get_props(), schema);
+                    auto res = RowReader::getPropByName(reader.get(), column);
+                    if (!ok(res)) {
+                        LOG(ERROR) << "Skip the bad value for prop " << column;
+                        continue;
+                    }
+                    v = value(std::move(res));
                 }
-                auto&& v = value(std::move(res));
+                yieldRecord.emplace_back(v);
                 record.emplace_back(std::move(v));
             }
+            // Check if duplicate, only for user-specified columns within yield clause
+            if (distinct_) {
+                auto ret = uniqResult->emplace(boost::hash_range(yieldRecord.begin(),
```

Comment:
Emm, the `DISTINCT` applies to a row, so it is better using `record` here.

Reply:
> Emm, the `DISTINCT` applies to a row, so it is better using `record` here.

The sentence proposed by this PR is `LOOKUP ON {<vertex_tag> | <edge_type>} WHERE YIELD DISTINCT <return_list>`. In my opinion, **DISTINCT** is more reasonable to play a role in the columns in the user-specified `<return_list>`. In fact, taking the LOOKUP statement for edges as an example, the columns `SrcVID`, `DstVID`, and `Ranking` are added by default in each row of the output. In this way, all the columns of each record will be different, and **DISTINCT** will not play the expected role.